deployment/aws-waf-security-automations.template

```diff
@@ -1967,6 +2035,33 @@
         },
         "CreateUniqueID": "true"
       }
+    },
+    "BlockedIpSNSTopic": {
+      "Type" : "AWS::SNS::Topic",
+      "Condition": "BlockedIpSNSTopicCreated",
+      "Properties" : {
+        "DisplayName": {
+          "Fn::Join": [" - ", [{
+            "Ref": "AWS::StackName"
+          }, "BlockedIpSNSTopic"]]
+        }
+      }
+    },
+    "BlockedIpSNSSubscription": {

```

Comment:
Not sure about this. The waf template creates the SNS topic but shouldn't care about any subscriptions. It kind of contradicts with the whole provider/subscriber idea? @michelesr what do you think?

Reply:
Yes, I thought the same - and happy to be argued either way. To note that this gives you an option (but no obligation - the default is not to have one) to create a subscription to the SNS at the point where you create an actual waf. Arguably without at least one subscription, having an sns doesn't achieve very much - and it's absolutely neutral as to the type of subscription - they all have exactly the same format: protocol + endpoint (e.g. staging-dms currently uses an e-mail protocol for pagerduty, but there are lots of other choices, e.g. protocol could be 'lambda' and endpoint a lambda arn).

Alternatively, for better separation, you could have an 'sns tier' for each product/environmen. All the subscriptions etc go in the sns tier, and you pass only the topic arn of the sns tier as a parameter to the waf template. Bonus is that you could use the same sns instance for any other alerts that you wanted to raise on a specific product (e.g. live stuff from cloudwatch).

Reply:
I think the template shouldn't offer to create subscription; it can create only one anyway..
It is best to let applications subscribe as needed (or don't subscribe at all) without having to change the waf template.

One more reason for the above is that adding the subscription resource parameters could introduce cyclic dependency between the tiers

Reply:
So sns tier then for those products that want to use sns for waf (or other monitoring), and pass its topic arn to the waf? Only hard-coded interdependency is that waf alerts will only go to an sns rather than other notification channel. But waf doesn't need to worry about how that sns is managed. I think if waf creates sns and application subscribes you could get circ dependency as waf wd rely on application (for subscription) when creating log-parser and application wd rely on waf (for sns topic arn) when creating subscription?

Reply:
why would waf depend on the app for subscription? Waf template creates the topic and exposes the topic arn in its outputs. Apps then subscribe as and when they wish. You will get

```yaml
app:
  params:
     topic: ${stack_output waf/topicArn}
     SubscrProtocol: email
     SubscrEndpoint: me@hotmail.com
```

```

Reply:
Ooops - my bad. Was thinking waf needed to know subscription for alerts, but it publishes to the topic arn.

So we'd need to amend the eb_environment template to do the subscriptions:
<img width="821" alt="screen shot 2018-07-18 at 16 39 30" src="https://user-images.githubusercontent.com/29107845/42892362-8d2ab100-8aa9-11e8-88f7-f11b4671449f.png">
Or does it vary by product (not done anything with app tier up to now!)?

Reply:
`app` can be anything really, I used the name for the sake of the example. In reality I think we have to have a notifications tier or similar which uses a template to create one or more subscriptions. This can be per app or even per environment as it would only require a topic arn

Reply:
Think that would be simplest. Given we're going to have one topic arn per app (since 1-1 reln between app and waf and each waf creates its own topic) presumably we need one notification tier per app too, all using an identical template? But I may be misunderstanding the terminology!

Reply:
For now, I think we just remove all the subscription stuff from waf template and get the new waf solution deployed to non-prod envs, and logging underway. Can create notification tiers separately at a later date as they won't affect anything upstream?

Reply:
Have pushed changes.